src/librustc_typeck/astconv.rs

```diff
@@ -970,60 +970,75 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx> + 'o {
     {
         let tcx = self.tcx();
 
-        if trait_bounds.is_empty() {
-            span_err!(tcx.sess, span, E0224,
-                      "at least one non-builtin trait is required for an object type");
-            return tcx.types.err;
-        }
-
         let mut projection_bounds = Vec::new();
+        let mut potential_assoc_types = Vec::new();
         let dummy_self = self.tcx().types.trait_object_dummy_self;
-        let (principal, potential_assoc_types) = self.instantiate_poly_trait_ref(
-            &trait_bounds[0],
-            dummy_self,
-            &mut projection_bounds,
-        );
-        debug!("principal: {:?}", principal);
-
-        for trait_bound in trait_bounds[1..].iter() {
-            // sanity check for non-principal trait bounds
-            self.instantiate_poly_trait_ref(trait_bound,
-                                            dummy_self,
-                                            &mut vec![]);
+        let bound_trait_refs: Vec<_> = trait_bounds
+            .iter()
+            .rev()
+            .map(|trait_bound| {
+                let (trait_ref, cur_potential_assoc_types) = self.instantiate_poly_trait_ref(
+                    trait_bound,
+                    dummy_self,
+                    &mut projection_bounds
+                );
+                potential_assoc_types.extend(cur_potential_assoc_types.into_iter().flatten());
+                (trait_ref, trait_bound.span)
+            }).collect();
+
+        // Expand trait aliases recursively and check that only one regular (non-auto) trait
+        // is used and no 'maybe' bounds are used.
+        let expanded_traits = traits::expand_trait_aliases(tcx, bound_trait_refs.clone());
```

Comment:
you're cloning a vector here, then `expand_trait_aliases` does `into_iter` on it and collects back into a `Vec`. The other use of `bound_trait_refs` also just calls `into_iter`. Maybe don't collect for `bound_trait_refs` (just build the iterator), have `expand_trait_aliases` take `impl Iterator` and just clone the iterator you pass in. The second use of `bound_trait_refs` can then consume the iterator.

Reply:
The problem is the mutable borrows of `projection_bounds` and `potential_assoc_types`, which means the iterator can't be cloned. I did want to do the same...

Reply:
I guess we can do this later, it would change the code too much. Please leave a FIXME

Reply:
Sure.

Reply:
`expand_trait_aliases` could take an `impl Iterator`, and you could just call `bound_trait_refs.iter()` here. That would get rid of another intermediate `Vec`

Reply:
This has not been addressed

Reply:
You can still collect into a vector above. What I mean is that you should not be cloning the vector here, but just call `.iter()` and make the `expand_trait_aliases` function take an `Iterator` instead of an `IntoIterator`.

Reply:
Oh, I see.